### DISCUSSION TITLE
Unbound variable when narrowing class attributes

### DISCUSSION THREAD

**Discussion by Pitrified** [microsoft/pyright#5399] [Q&A] [answered]
Hello, and thank you for your support

If I use a `Literal` type in a `if` statement in a function, everything works as I expect:

```python
from typing import Literal
ENV_TYPE_LIT = Literal["dev", "prod"]

def sample_func(env: ENV_TYPE_LIT) -> None:
    if env == "dev":  #    (parameter) env: ENV_TYPE_LIT
        a = 1
    elif env == "prod":  # (parameter) env: Literal['prod']
        a = 2

    print(a)  # (variable) a: Literal[1, 2]
```

However when I use the same pattern inside a class, `a` is reported as possibly unbound.

```python
class Sample:
    def __init__(self, env: ENV_TYPE_LIT) -> None:
        self.env: ENV_TYPE_LIT = env

        if self.env == "dev":  #    (variable) env: ENV_TYPE_LIT
            a = 1
        elif self.env == "prod":  # (variable) env: Literal['prod']
            a = 2
        elif self.env == "....":  # (variable) env: Never
            pass

        print(a)  # (variable) a: Unbound | Literal[1, 2]
```

I see that the type is being narrowed, and after the two options are exhausted it is reported as `Never`, so I do not understand why it's happening.
I've looked around and found this issue regarding [literal class attributes](https://github.com/microsoft/pyright/issues/4828#issuecomment-1481109030), but explicitly setting `self.env: ENV_TYPE_LIT` did not help (we are in the constructor anyway).

How should I fix this problem?

I'm even more confused about why this code also leads to `b` being possibly unbound:

```python
class Sample:
    def __init__(self, env: ENV_TYPE_LIT) -> None:
        if env == "dev":  #    (parameter) env: ENV_TYPE_LIT
            b = 1
        elif env == "prod":  # (parameter) env: Literal['prod']
            b = 2
        elif env == "....":  # (parameter) env: Never
            pass
        print(b)  # (variable) b: Unbound | Literal[1, 2]
```

How is this snippet different from the first one using `sample_func`?

Thank you very much for your help,
Peter

**Comment by erictraut** [comment, marked as answer]
In your code sample, you're leveraging a pyright feature called "implied else type narrowing". It works in certain situations where an "if" or "elif" statement has no corresponding "else". It allows pyright to determine (under certain circumstances) that the "implied else" clause is never "executed" because of type exhaustion. This is a very advanced type checking feature -- one that is not supported by mypy or other Python type checkers. When I implemented it initially, we noticed an immediate and significant performance impact for all type analysis. I found a way to mitigate the performance impact but only by imposing certain restrictions on the feature. For more information about these restrictions, search for "implied else" in the pyright documentation. One of these restrictions is that it supports type narrowing only for simple names (like "env"), not for member access expressions (like "self.env").

If you're interested in why this feature is difficult to implement... Normally, the code flow graph influences type narrowing. In this case, type narrowing effectively modifies the code flow graph (by adding an "implied else" code block whose code block is unreachable). This circularity (where type narrowing affects the code graph which affects code narrowing) is difficult to handle.

I can offer a couple of workarounds. The first is to use a local variable for your type guards.

```python
class Sample:
    def __init__(self, env: ENV_TYPE_LIT) -> None:
        self.env: ENV_TYPE_LIT = env

        env_temp: ENV_TYPE_LIT = self.env

        if env_temp == "dev":
            a = 1
        elif env_temp == "prod":
            a = 2

        print(a)
```

The second is to use an explicit `else` block with an `assert_never` call. This is the approach I would personally use in this case. It's also the approach that mypy users would typically use for all "implied else" use cases.

```python
class Sample:
    def __init__(self, env: ENV_TYPE_LIT) -> None:
        self.env: ENV_TYPE_LIT = env

        if self.env == "dev":
            a = 1
        elif self.env == "prod":
            a = 2
        else:
            assert_never(self.env)

        print(a)
```

I'm going to close this issue because pyright is working as intended here.

**Reply by Pitrified** [reply]
Fantastic, thank you for the explanation and the solutions!
